### PR TITLE
Allow package.json "main" to specify a directory

### LINKED ISSUE #12198: node moduleResolver does not handle directory in package.json main field
The `common-tags` package indicates a directory in its `main` field in the `package.json`, which is valid in node (it will look in `lib/index.js`, `lib/package.json`, etc).

The entry can be seen [here](https://github.com/declandewet/common-tags/blob/master/package.json#L84).

This a snippet of the output of `--traceResolution` in a project:
```
File '.../src/node_modules/common-tags.js' does not exist.
File '.../src/node_modules/common-tags.jsx' does not exist.
Found 'package.json' at '.../src/node_modules/common-tags/package.json'.
No types specified in 'package.json', so returning 'main' value of 'lib'
File '.../src/node_modules/common-tags/lib' does not exist.
File '.../src/node_modules/common-tags/lib.ts' does not exist.
File '.../src/node_modules/common-tags/lib.tsx' does not exist.
File '.../src/node_modules/common-tags/lib.d.ts' does not exist.
File '.../src/node_modules/common-tags/index.js' does not exist.
File '.../src/node_modules/common-tags/index.jsx' does not exist.
File '.../src/node_modules/@types/common-tags.js' does not exist.
File '.../src/node_modules/@types/common-tags.jsx' does not exist.
File '.../src/node_modules/@types/common-tags/package.json' does not exist.
File '.../src/node_modules/@types/common-tags/index.js' does not exist.
File '.../src/node_modules/@types/common-tags/index.jsx' does not exist.
File '.../node_modules/common-tags.js' does not exist.
```

The resolver sees the `main` field correctly, but it doesn't try accessing it as a directory, only as a file; eventually failing to locate the package.

Using typescript ea309fe5.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #12198

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2877,10 +2877,6 @@
         "category": "Message",
         "code": 6136
     },
-    "No types specified in 'package.json', so returning 'main' value of '{0}'": {
```
Comment (andy-ms) [inline src/compiler/diagnosticMessages.json]: This message is not needed any more - now we do the full TS lookup before doing the JS lookup.

File: src/compiler/moduleNameResolver.ts [hunk 2]
```diff
@@ -670,45 +661,55 @@ namespace ts {
...
+        if (file) {
+            const onlyRecordFailures = !directoryProbablyExists(getDirectoryPath(file), state.host);
+            const fromFile = tryFile(file, failedLookupLocations, onlyRecordFailures, state);
+            if (fromFile) {
+                // Note: this would allow a package.json to specify a ".js" file as typings. Maybe that should be forbidden.
+                return resolvedFromAnyFile(fromFile);
+            }
+
+            if (ts) {
+                //should this be potentially a directory?
```
Comment (andy-ms) [inline src/compiler/moduleNameResolver.ts]: @vladima

**Comment by andy-ms** [comment]
Waiting for nodejs/node#9642 before proceeding.

**Comment by mhegazy** [comment]
@andy-ms can you refresh this change?

**Comment by andy-ms** [comment]
This has gotten out of date, so redoing as #13678.

**Pull request closed without merging**